src/turtle/env/model/TestNode.d

```diff
@@ -0,0 +1,249 @@
+    import ocean.net.server.connection.IConnectionHandlerInfo;
+    import ocean.io.select.protocol.generic.ErrnoIOException;
+    import ocean.util.log.Logger;
+
+    /// Enum defining the possibles states of the fake node service.
+    private enum State
+    {
+        Init,
+        Running,
+        Stopped
+    }
+
+    /// State of the fake node service.
+    private State state;
+
+    /// Used to prevent creating multiple fake nodes of the same type.
+    static bool already_created = false;
+
+    /// Flag indicating that unhandled exceptions from the node must be printed
+    /// in test suite trace.
+    private bool log_errors = true;
+
+    /// Logger shared by all turtle nodes of this type.
+    static private Logger log;
+
+    /***************************************************************************
+
+        Static constructor.
+
+    ***************************************************************************/
+
+    static this ( )
+    {
+        log = Log.lookup("fakedmq");
+    }
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            node = node addres & port
+            neo_port = port of neo listener (same address as above)
+            conn_setup_params = connection handler constructor arguments
+            options = options for the neo node and connection handlers
+            backlog = (see ISelectListener ctor)
+
+    ***************************************************************************/
+
+    public this ( AddrPort node, ushort neo_port,
+        ConnectionSetupParams conn_setup_params, Options options, int backlog )
+    {
+        verify(!already_created, "Can only have one " ~ this.id ~ " per turtle test app");
+        already_created = true;
+
+        super(node, neo_port, conn_setup_params, options, backlog);
+        this.error_callback = &this.onError;
+    }
+
+    /***************************************************************************
+
+        Starts the fake node as part of test suite event loop. It will
+        only terminate when whole test suite process dies.
+
+    ***************************************************************************/
+
+    public void start ( )
+    {
+        verify(this.state == State.Init, "Node has already been started");
+        this.state = State.Running;
+        turtle_env_registry.register(this);
+
+        this.register(theScheduler.epoll);
+        theScheduler.processEvents();
+    }
+
+    /***************************************************************************
+
+        Restarts the fake node, reopening the listening socket on the same port
+        determined in the initial call to start().
+
+        Notes:
+            1. Restarting the node *does not* clear any data in its storage
+               engine. To do that, call clear().
+            2. You must call stop() first, before calling restart().
```

Comment:
This is extremely weird tbh. If you need to call `stop` first, it is just `start`, not `restart`.

Reply:
Agreed. I have no memory of why we did it like this.